### PR TITLE
feat: support assistant settings and implement structlog

### LINKED ISSUE #19: [FEATURE] Use the Assistant Context if it is not provided during runtime
Problem:
When creating an Assistant, we can provide the config/context to the graph. But when we try to use an assistant in like Agent Chat App, we cannot provide `config/context` for very /run/stream API call. So, if there is a config/context attached to Assistant during it's creation, it should be leveraged. 

Solution:
A simple `if` block to assign the assistants context during a `run`.

- API changes needed

- [ ] Low - Nice to have

### PR DESCRIPTION
**Pull request by arthiondaena**
## Type of Change

- [x] `feat`: New feature
- [ ] `fix`: Bug fix
- [ ] `docs`: Documentation changes
- [ ] `style`: Code style/formatting
- [x] `refactor`: Code refactoring
- [ ] `perf`: Performance improvement
- [ ] `test`: Tests added/updated
- [ ] `chore`: Maintenance (dependencies, build, etc.)
- [ ] `ci`: CI/CD changes

## Related Issues

Related #39 
Closes #19 

## Changes Made

- Fixed all the issues related to agent settings in langsmith, which includes
  - Versioning for default assistants
  - Flags required by langsmith to access assistant settings. (`/info` endpoint)
  - Use assistant config in thread as default config if config/context is not explicitly sent for run creation.
- Replace logging module with structlog, which is more suitable for production environment.x

## Testing

- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [ ] E2E tests added/updated
- [ ] Manual testing performed

## Checklist

- [x] Code follows project style (Ruff formatting)
- [x] All linting checks pass (`make lint`)
- [x] Type checking passes (`make type-check`)
- [x] All tests pass (`make test`)
- [x] Documentation updated (if needed)
- [x] Commit messages follow [Conventional Commits](https://www.conventionalcommits.org/)
- [x] PR title follows Conventional Commits format

### CODE REVIEW

File: src/agent_server/api/runs.py [hunk 1]
```diff
@@ -145,6 +154,19 @@ async def create_run(
...
+        raise HTTPException(
+            status_code=400,
+            detail="Cannot specify both configurable and context. Prefer setting context alone. Context was introduced in LangGraph 0.6.0 and is the long term planned replacement for configurable.",
+        )
+
+    if context:
+        configurable = context.copy()
+        config["configurable"] = configurable
+    else:
+        context = configurable.copy()
```
Comment (ibbybuilds) [inline src/agent_server/api/runs.py:160]: what if context schema is different from configurable? because the reason for removing it was that on langgraph sutdio the schema for context was not matching (there was thread_id coming in config which was not present in context schema). have you tested this with langgraph studio?

File: src/agent_server/api/runs.py [hunk 2]
```diff
@@ -104,6 +104,15 @@ async def update_thread_metadata(
     await session.commit()
 
 
+def _merge_jsonb(*objects: dict) -> dict:
```
Comment (ibbybuilds) [inline src/agent_server/api/runs.py]: better to move the method as utility in utils file

File: src/agent_server/core/health.py [hunk 3]
```diff
@@ -35,6 +36,12 @@ async def info() -> InfoResponse:
         version="0.1.0",
         description="Production-ready Agent Protocol server built on LangGraph",
         status="running",
+        flags={
+            "assistants": True,
+            "crons": False,
+            "langsmith": True,
```
Comment (ibbybuilds) [inline src/agent_server/core/health.py]: why is langsmith true here, we don't support it yet i think?

File: src/agent_server/core/health.py [hunk 4]
```diff
@@ -35,6 +36,12 @@ async def info() -> InfoResponse:
...
+            "langsmith_tracing_replicas": True,
```
Comment (ibbybuilds) [inline src/agent_server/core/health.py]: same here

File: run_server.py [hunk 5]
```diff
@@ -68,17 +75,18 @@ def main():
...
-    print("🧪 Test with: python test_sdk_integration.py")
+    logger.info("🚀 Starting Aegra...")
+    logger.info(f"📍 Server will be available at: http://localhost:{port}")
+    logger.info(f"📊 API docs will be available at: http://localhost:{port}/docs")
+    logger.info("🧪 Test with: python test_sdk_integration.py")
 
     uvicorn.run(
         "src.agent_server.main:app",
-        host="0.0.0.0",
+        host=os.getenv("HOST", "127.0.0.1"),
```
Comment (ibbybuilds) [inline run_server.py]: **Problem:**
Changing the default host from `0.0.0.0` to `127.0.0.1` breaks Docker containers and production deployments. Containers need `0.0.0.0` to accept external connections.

**Current code:**
```python
host=os.getenv("HOST", "127.0.0.1"),
```

**Solution:**
```python
host=os.getenv("HOST", "0.0.0.0"),  # Restore original default
```

This maintains backward compatibility while still allowing override via env var.

File: src/agent_server/main.py [hunk 6]
```diff
@@ -20,8 +20,8 @@
     sys.path.insert(0, str(graphs_dir))
 
 # ruff: noqa: E402 - imports below require sys.path modification above
```
Comment (ibbybuilds) [inline src/agent_server/main.py:22]: **Problem:**
`setup_logging()` is only called in `run_server.py`, but not in `main.py`. When running the app directly via uvicorn (e.g., `uvicorn src.agent_server.main:app`), structlog is never configured, causing logging to fail or use wrong format.

**Current code:**
```python
import structlog
logger = structlog.getLogger(__name__)
```

**Solution:**
```python
import structlog
from .utils.setup_logging import setup_logging

# Configure logging at module level
setup_logging()
logger = structlog.getLogger(__name__)
```

This ensures logging works regardless of entry point.

File: src/agent_server/utils/setup_logging.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
+import logging
+import logging.config
+import os
+from typing import Any
+
+import structlog
+
+
+class InterceptHandler(logging.Handler):
```
Comment (ibbybuilds) [inline src/agent_server/utils/setup_logging.py]: ### ⚠️ Dead code: InterceptHandler is never used

**Problem:**
This handler is defined but never used anywhere. Additionally, the implementation just calls `logger.handle(record)` on the same record, which could cause infinite recursion.

**Current code:**
```python
class InterceptHandler(logging.Handler):
    def emit(self, record: logging.LogRecord) -> None:
        logger = logging.getLogger(record.name)
        logger.handle(record)
```
**Solution:**
Either remove it entirely, or implement it properly to bridge stdlib → structlog:

```python
class InterceptHandler(logging.Handler):
    """Intercept standard logging and forward to structlog"""
    def emit(self, record: logging.LogRecord) -> None:
        logger = structlog.get_logger(record.name)
        level = record.levelname.lower()
        getattr(logger, level)(record.getMessage(), exc_info=record.exc_info)
```

File: src/agent_server/api/runs.py (hunk 1 above)
Comment (arthiondaena) [reply src/agent_server/api/runs.py:160]: This is exactly how the original langgraph is implemented.
And yes, langgraph studio perfectly works with this implementation. You can try to test it once, and let me know if you encounter any error.
In our case context schema will differ from configurable, since we will be injecting thread_id and other values. Context schema must only have key/value pairs which are defined in Context class ([`graphs/react_agent/context.py`](https://github.com/ibbybuilds/aegra/blob/1bbaf808bcb628259d4d8756e5d6c3537fca3d36/graphs/react_agent/context.py#L13))

File: src/agent_server/api/runs.py (hunk 2 above)
Comment (arthiondaena) [reply src/agent_server/api/runs.py]: Got it

File: run_server.py (hunk 5 above)
Comment (arthiondaena) [reply run_server.py]: The reason for this was because of pre-commit hook. It's getting failed because of hardcoded bind to all interfaces.
```
check for merge conflicts................................................Passed
bandit...................................................................Failed
- hook id: bandit
- exit code: 1

[main]  INFO    profile include tests: None
[main]  INFO    profile exclude tests: B101
[main]  INFO    cli include tests: None
[main]  INFO    cli exclude tests: None
[main]  INFO    using config: pyproject.toml
[main]  INFO    running on Python 3.11.13
Run started:2025-10-14 14:58:54.868663

Test results:
>> Issue: [B104:hardcoded_bind_all_interfaces] Possible binding to all interfaces.
   Severity: Medium   Confidence: Medium
   CWE: CWE-605 (https://cwe.mitre.org/data/definitions/605.html)
   More Info: https://bandit.readthedocs.io/en/0.0.0/plugins/b104_hardcoded_bind_all_interfaces.html
   Location: .\run_server.py:85:31
84              "src.agent_server.main:app",
85              host=os.getenv("HOST", "0.0.0.0"),
86              port=port,

--------------------------------------------------

Code scanned:
        Total lines of code: 66
        Total lines skipped (#nosec): 0
        Total potential issues skipped due to specifically being disabled (e.g., #nosec BXXX): 0

Run metrics:
        Total issues (by severity):
                Undefined: 0
                Low: 0
                Medium: 1
                High: 0
        Total issues (by confidence):
                Undefined: 0
                Low: 0
                Medium: 1
                High: 0
Files skipped (0):
```

Comment (arthiondaena) [reply run_server.py]: Should I disable ruff for this line?

Comment (ibbybuilds) [reply run_server.py]: `# nosec B104 - required for Docker`

yes add this

File: src/agent_server/utils/run_utils.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import copy
+from typing import Any
+
+
+def _should_skip_event(raw_event: Any) -> bool:
```
Comment (ibbybuilds) [inline src/agent_server/utils/run_utils.py:5]: these are not private methods anymore, please remove the underscrore prefix
